GenericList/string_list.cpp

```diff
@@ -56,50 +102,86 @@ std::string List_String::ToString() {
 	return result;
 }
 
-
-bool List_String::Add(std::string new_val) {
-	if (capacity < count + 1) {
-		Capacity(capacity == 0 ? 1 : capacity * 2); //double array size
+std::string* List_String::Find(const std::string& val) {
+	for (int i = 0; i < count; i++) {
+		if (data[i] == val) return &data[i];
 	}
-	data[count++] = new_val;
-	return true;
+	return nullptr;
 }
 
-bool List_String::Contains(std::string val) {
-	return IndexOf(val) != -1;
+const std::string* List_String::Find(const std::string& val) const {
+	for (int i = 0; i < count; i++) {
+		if (data[i] == val) return &data[i];
+	}
+	return nullptr;
 }
 
-bool List_String::RemoveAt(int index) {
-	if (index < 0) return false;
-	if (index >= count) return false;
-	//allowed setting: index = [0, count-1], count > 0 (data is initialized)
-
-	std::memmove(data + index, data + index + 1, count - index - 1);
-	count--;
-	return true;
+void List_String::RemoveAt(int index) {
+	if (index < 0 || index >= count) throw std::out_of_range(string("Cannot remove element at out_of_range index: ") + to_string(index) + string("."));
+	
+	std::move(data + index + 1, data + count, data + index);
+	data[--count].~string();
 }
 
 
-int List_String::IndexOf(std::string val) {
-	for (int i = 0; i < count; i++) {
-		if (data[i] == val) return i;
-	}
-	return -1;
+
+bool List_String::Remove(const std::string& val) {
+	std::string* ptr = Find(val);
+	if (ptr!=nullptr) RemoveAt(ptr-data);
+	return ptr!=nullptr;
 }
 
 
-bool List_String::Remove(std::string val) {
-	int index = IndexOf(val);
-	return RemoveAt(index);
+const std::string& List_String::Get(int index) const {
+	if (index >= count || index < 0) throw std::out_of_range(string("Cannot get element at out_of_range index: ") + to_string(index) + string(")"));
+
+	return data[index];
 }
 
-std::string List_String::Get(int index) {
+std::string& List_String::Get(int index) {
 	if (index >= count || index < 0) throw std::out_of_range(string("Cannot get element at out_of_range index: ") + to_string(index) + string(")"));
 
 	return data[index];
 }
 
 
+std::string* List_String::CreateDeepCopy(std::string* data, size_t data_size, size_t copy_size) {
+	if (copy_size > data_size || copy_size < 0) {
+		throw std::out_of_range(string("Cannot copy array of size ") + to_string(copy_size) + string(" onto array of size ") + to_string(copy_size) + string("."));
+	}
+	std::string* new_data = data_size > 0 ? dataAllocator.allocate(data_size) : nullptr;
+
+
+	//For improved performance, replace copy by memcpy+fill (no deep copy of non-POD objects) or by using a swap method
+	for (size_t i = 0; i < copy_size; i++) {
+		new (new_data + i) std::string(data[i]);
+	}
+
+	return new_data;
+}
+
+void List_String::Resize(int new_capacity) {
+	if (new_capacity < count) return; //no change
```

Comment:
Since this is a private function that is only called after you've made the check elsewhere, you can just `assert`, on release builds the `assert` gets removed
```suggestion
	assert(new_capacity > count);
```